pkg/pgclient/config.go

```diff
@@ -29,30 +29,52 @@ type Config struct {
+	flag.StringVar(&cfg.Host, "db-host", DefaultDBHost, "Host for TimescaleDB/Vanilla Postgres.")
+	flag.IntVar(&cfg.Port, "db-port", DefaultDBPort, "TimescaleDB/Vanilla Postgres connection password.")
+	flag.StringVar(&cfg.User, "db-user", DefaultDBUser, "TimescaleDB/Vanilla Postgres user.")
+	flag.StringVar(&cfg.Password, "db-password", DefaultDBPassword, "Password for connecting to TimescaleDB/Vanilla Postgres.")
+	flag.StringVar(&cfg.Database, "db-name", DefaultDBName, "Database name.")
+	flag.StringVar(&cfg.SslMode, "db-ssl-mode", DefaultSSLMode, "TimescaleDB/Vanilla Postgres connection ssl mode. If you do not want to use ssl, pass 'allow' as value.")
 	flag.IntVar(&cfg.DbConnectRetries, "db-connect-retries", 0, "Number of retries Promscale should make for establishing connection with the database.")
 	flag.BoolVar(&cfg.AsyncAcks, "async-acks", false, "Acknowledge asynchronous inserts. If this is true, the inserter will not wait after insertion of metric data in the database. This increases throughput at the cost of a small chance of data loss.")
 	flag.IntVar(&cfg.ReportInterval, "tput-report", 0, "Interval in seconds at which throughput should be reported.")
 	flag.Uint64Var(&cfg.LabelsCacheSize, "labels-cache-size", 10000, "Maximum number of labels to cache.")
 	flag.Uint64Var(&cfg.MetricsCacheSize, "metrics-cache-size", pgmodel.DefaultMetricCacheSize, "Maximum number of metric names to cache.")
 	flag.IntVar(&cfg.WriteConnectionsPerProc, "db-writer-connection-concurrency", 4, "Maximum number of database connections for writing per go process.")
 	flag.IntVar(&cfg.MaxConnections, "db-connections-max", -1, "Maximum number of connections to the database that should be opened at once. It defaults to 80% of the maximum connections that the database can handle.")
+	flag.StringVar(&cfg.DbUri, "db-uri", DefaultDBUri, "TimescaleDB/Vanilla Postgres DB URI. Example DB URI `postgres://postgres:password@localhost:5432/timescale?sslmode=require`")
 	return cfg
 }
 
+var excessDBFlagsError = fmt.Errorf("failed to build DB credentials with provided flags. Please use either db flags or db-uri not both")
+
 // GetConnectionStr returns a Postgres connection string
-func (cfg *Config) GetConnectionStr() string {
-	return fmt.Sprintf("host=%v port=%v user=%v dbname=%v password='%v' sslmode=%v connect_timeout=10",
-		cfg.Host, cfg.Port, cfg.User, cfg.Database, cfg.Password, cfg.SslMode)
+func (cfg *Config) GetConnectionStr() (string, error) {
+	if cfg.DbUri == DefaultDBUri {
+		return fmt.Sprintf("host=%v port=%v user=%v dbname=%v password='%v' sslmode=%v connect_timeout=10",
+			cfg.Host, cfg.Port, cfg.User, cfg.Database, cfg.Password, cfg.SslMode), nil
+	} else if cfg.Database != DefaultDBName || cfg.Host != DefaultDBHost || cfg.Port != DefaultDBPort || cfg.User != DefaultDBUser ||
+		cfg.Password != DefaultDBPassword || cfg.SslMode != DefaultSSLMode {
+		return "", excessDBFlagsError
+	}
+
+	uri := cfg.DbUri + "&connect_timeout=10"
+	return uri, nil
```

Comment:
```suggestion
func (cfg *Config) GetConnectionStr() (string, error) {
    if cfg.Database != DefaultDBName ||
        cfg.Host != DefaultDBHost ||
        cfg.Port != DefaultDBPort ||
        cfg.User != DefaultDBUser ||
        cfg.Password != DefaultDBPassword ||
        cfg.SslMode != DefaultSSLMode {

        if cfg.DbUri != DefaultDBUri {
            return "", excessDBFlagsError
        }
        return fmt.Sprintf("host=%v port=%v user=%v dbname=%v password='%v' sslmode=%v connect_timeout=10",
            cfg.Host, cfg.Port, cfg.User, cfg.Database, cfg.Password, cfg.SslMode), nil
    }
    return cfg.DbUri + "&connect_timeout=10", nil
```

Reply:
This needs some unit tests with the increased complexity.

Reply:
Yeah, I will add unit tests for this. But the code you suggested with fail in one edge case. i.e if user wants to connect to DB with the default flags, in that case, nothing will be changed with the above validation we cannot get into the line 78 build `("host=%v port=%v....)` this case will return by default `   &connect_timeout=10` which isn't expected.